tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -245,98 +242,171 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
+    transitive_deps = []
+    # We care about the order of dependencies list, so it will be a list.
+    collapsed_deps = []
+    # We don't care about the order of srcs and headers, they will be sorted.
+    collapsed_srcs = set(_extract_sources(bazel_rule))
+    collapsed_public_headers = set(_extract_public_headers(bazel_rule))
+    collapsed_headers = set(_extract_nonpublic_headers(bazel_rule))
+    exclude_deps = set()
+
+    for dep in direct_deps:
+        external_dep_name_maybe = _external_dep_name_from_bazel_dependency(dep)
+
+        if dep in bazel_rules:
+            # The follow if is used to suppress expanding absl libs; expanding
+            # absl deps are technically correct but creates too many noise
+            if external_dep_name_maybe is None or not external_dep_name_maybe.startswith(
+                    'absl'):
+                if "_PROCESSING_DONE" not in bazel_rules[dep]:
+                    # This item is not processed before, compute now
+                    _compute_transitive_metadata(dep, bazel_rules,
+                                                 bazel_label_to_dep_name)
+                _deduplicate_append(
+                    transitive_deps,
+                    bazel_rules[dep].get('_TRANSITIVE_DEPS', []))
+                _deduplicate_append(collapsed_deps,
+                                    bazel_rules[dep].get('_COLLAPSED_DEPS', []))
+                exclude_deps.update(bazel_rules[dep].get('_EXCLUDE_DEPS', []))
+
+        # This dep is a public target, add it as a dependency
+        if dep in bazel_label_to_dep_name:
+            _deduplicate_append(transitive_deps, [bazel_label_to_dep_name[dep]])
+            _deduplicate_append(collapsed_deps, [bazel_label_to_dep_name[dep]])
+            exclude_deps.update(bazel_rules[dep]['_TRANSITIVE_DEPS'])
+            exclude_deps.add(dep)
+            continue
 
-    expansion_blocklist = set()
-    to_expand = set(bazel_deps)
-    while to_expand:
+        # This dep is an external target, add it as a dependency
+        if external_dep_name_maybe is not None:
+            _deduplicate_append(transitive_deps, [external_dep_name_maybe])
+            _deduplicate_append(collapsed_deps, [external_dep_name_maybe])
+            continue
 
-        # start with the last dependency to be built
-        build_order = _sort_by_build_order(list(to_expand), bazel_rules,
-                                           'transitive_deps')
+    # Direct dependencies are part of transitive dependencies
+    _deduplicate_append(transitive_deps, direct_deps)
+
+    # Removes all duplicated intermediate labels.
+    # This is the step that further shorten the deps list.
+    collapsed_deps = list(
+        filter(lambda x: x not in exclude_deps, collapsed_deps))
+
+    # Compute the final source files and headers, without all intermediate dependencies.
+    for dep in transitive_deps:
+        if dep not in exclude_deps:
```

Comment:
I think that example makes sense with two things:
- if you take care to explain what "excluded_deps" actually means (I think it means "all transitive deps of public deps of a give target") this would be so much easier to digest.
- The  example is hard to read since the  letters ABCDE are assigned to targets randomly (so you have to read it multiple times to remember which targets are public, which are intemediate etc.).  If you marked the targets e.g. "PA", "PB", "PC" for public targets and "IX, IY, IZ" for intermediate targets, digesting it would be easier.

Reply:
Updated as suggested:

```
    # Imaging a public target PX has transitive deps [IA, IB, PY, IC, PZ]. PX,
    # PY and PZ are public build targets. And IA, IB, IC are intermediate
    # targets. In addition, PY depends on IC.
    #
    # Translate the condition into dependency graph:
    #   PX -> [IA, IB, PY, IC, PZ]
    #   PY -> [IC]
    #   Public targets: [PX, PY, PZ]
    #
    # The collapsed dependencies of PX: [PY, PZ].
    # The excluded dependencies of X: [PY, IC, PZ].
    # (IC is excluded as a dependency of PX. It is already included in PY, hence
    # it would be redundant to include it again.)
    #
    # Target PX should include source files and headers of [PX, IA, IB] as final
    # build metadata.
```